File: src/main/java/cn/sd/jrz/jtool/date/JDate.java
package cn.sd.jrz.jtool.date;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Calendar;
import java.util.Date;
import java.util.Objects;

/**
 * @author JRZ
 * @since 1.0
 */
public final class JDate {
    private final Instant date;

    public static JDate of(String date) {
        return of(date, "yyyy-MM-dd HH:mm:ss");
    }

    public static JDate of(String date, String format) {
        SimpleDateFormat dateFormat = new SimpleDateFormat(format);
        Date d;
        try {
            d = dateFormat.parse(date);
        } catch (ParseException e) {
            throw new RuntimeException(e);
        }
        if (d == null) {
            return null;
        }
        return new JDate(d);
    }

    public static JDate of(int year, int month, int day, int hour, int minute, int second) {
        return new JDate(LocalDateTime.of(year, month, day, hour, minute, second));
    }

    public JDate() {
        this.date = new Date().toInstant();
    }

    public JDate(Date date) {
        this.date = date.toInstant();
    }

    public JDate(Calendar calendar) {
        this.date = calendar.getTime().toInstant();
    }

    public JDate(LocalDateTime localDateTime) {
        this.date = localDateTime.atZone(ZoneId.systemDefault()).toInstant();
    }

    public final Date toDate() {
        return Date.from(date);
    }

    public final LocalDateTime toLocalDateTime() {
        return LocalDateTime.ofInstant(date, ZoneId.systemDefault());
    }

    public final Calendar toCalendar() {
        Calendar instance = Calendar.getInstance();
        instance.setTime(toDate());
        return instance;
    }

    public final JDate addSeconds(long seconds) {
        return new JDate(toLocalDateTime().plusSeconds(seconds));
    }

    public final JDate addMinutes(long minutes) {
        return new JDate(toLocalDateTime().plusMinutes(minutes));
    }

    public final JDate addHours(long hours) {
        return new JDate(toLocalDateTime().plusHours(hours));
    }

    public final JDate addDays(long days) {
        return new JDate(toLocalDateTime().plusDays(days));
    }

    public final JDate addMonths(long months) {
        return new JDate(toLocalDateTime().plusMonths(months));
    }

    public final JDate addYears(long years) {
        return new JDate(toLocalDateTime().plusYears(years));
    }

    public final int second() {
        return toLocalDateTime().getSecond();
    }

    public final int minute() {
        return toLocalDateTime().getMinute();
    }

    public final int hour() {
        return toLocalDateTime().getHour();
    }

    public final int day() {
        return toLocalDateTime().getDayOfMonth();
    }

    public final int month() {
        return toLocalDateTime().getMonthValue();
    }

    public final int year() {
        return toLocalDateTime().getYear();
    }

    public final boolean isBefore(JDate jDate) {
        return toLocalDateTime().isBefore(jDate.toLocalDateTime());
    }

    public final boolean isBefore(Date date) {
        return isBefore(new JDate(date));
    }

    public final boolean isBefore(LocalDateTime localDateTime) {
        return isBefore(new JDate(localDateTime));
    }

    public final boolean isBefore(Calendar calendar) {
        return isBefore(new JDate(calendar));
    }

    public final boolean isAfter(JDate jDate) {
        return toLocalDateTime().isAfter(jDate.toLocalDateTime());
    }

    public final boolean isAfter(Date date) {
        return isAfter(new JDate(date));
    }

    public final boolean isAfter(LocalDateTime localDateTime) {
        return isAfter(new JDate(localDateTime));
    }

    public final boolean isAfter(Calendar calendar) {
        return isAfter(new JDate(calendar));
    }

    public final long currentTimeMillis() {
        return toDate().getTime();
    }

    @Override
    public final String toString() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        return dateFormat.format(toDate());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        JDate jDate = (JDate) o;
        return Objects.equals(date, jDate.date);
    }

    @Override
    public int hashCode() {
        return Objects.hash(date);
    }
}

File: src/main/java/cn/sd/jrz/jtool/date/JDate.kt
package cn.sd.jrz.jtool.date

import java.text.SimpleDateFormat
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*

/**
 * @author 江荣展
 * @date 2020/6/30
 */
class JDate(private val date: Instant) {

    companion object {
        fun of(date: String, format: String = "yyyy-MM-dd HH:mm:ss"): JDate? {
            return try {
                JDate(SimpleDateFormat(format).parse(date))
            } catch (e: Exception) {
                null
            }
        }

        fun of(year: Int, month: Int, day: Int, hour: Int, minute: Int, second: Int): JDate? {
            return JDate(LocalDateTime.of(year, month, day, hour, minute, second))
        }
    }

    constructor() : this(Date().toInstant())

    constructor(date: Date) : this(date.toInstant())

    constructor(calendar: Calendar) : this(calendar.time.toInstant())

    constructor(localDateTime: LocalDateTime) : this(localDateTime.atZone(ZoneId.systemDefault()).toInstant())

    fun toDate(): Date = Date.from(date)

    fun toCalendar(): Calendar = Calendar.getInstance().apply { time = toDate() }

    fun toLocalDateTime(): LocalDateTime = LocalDateTime.ofInstant(date, ZoneId.systemDefault())

    fun addSeconds(seconds: Long) = JDate(toLocalDateTime().plusSeconds(seconds))

    fun addMinutes(minutes: Long) = JDate(toLocalDateTime().plusMinutes(minutes))

    fun addHours(hours: Long) = JDate(toLocalDateTime().plusHours(hours))

    fun addDays(days: Long) = JDate(toLocalDateTime().plusDays(days))

    fun addMonths(months: Long) = JDate(toLocalDateTime().plusMonths(months))

    fun addYears(years: Long) = JDate(toLocalDateTime().plusYears(years))

    val second get() = toLocalDateTime().second

    val minute get() = toLocalDateTime().minute

    val hour get() = toLocalDateTime().hour

    val day get() = toLocalDateTime().dayOfMonth

    val month get() = toLocalDateTime().monthValue

    val year get() = toLocalDateTime().year

    fun isBefore(jDate: JDate) = toLocalDateTime().isBefore(jDate.toLocalDateTime())

    fun isBefore(date: Date) = isBefore(JDate(date))

    fun isBefore(localDateTime: LocalDateTime) = isBefore(JDate(localDateTime))

    fun isBefore(calendar: Calendar) = isBefore(JDate(calendar))

    fun isAfter(jDate: JDate) = toLocalDateTime().isAfter(jDate.toLocalDateTime())

    fun isAfter(date: Date) = isAfter(JDate(date))

    fun isAfter(localDateTime: LocalDateTime) = isAfter(JDate(localDateTime))

    fun isAfter(calendar: Calendar) = isAfter(JDate(calendar))

    operator fun compareTo(jDate: JDate): Int {
        return when {
            isBefore(jDate) -> -1
            isAfter(jDate) -> 1
            else -> 0
        }
    }

    operator fun compareTo(date: Date): Int {
        return when {
            isBefore(date) -> -1
            isAfter(date) -> 1
            else -> 0
        }
    }

    operator fun compareTo(localDateTime: LocalDateTime): Int {
        return when {
            isBefore(localDateTime) -> -1
            isAfter(localDateTime) -> 1
            else -> 0
        }
    }

    operator fun compareTo(calendar: Calendar): Int {
        return when {
            isBefore(calendar) -> -1
            isAfter(calendar) -> 1
            else -> 0
        }
    }

    val millis get() = toDate().time

    override fun toString(): String = SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(toDate())

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as JDate
        if (date != other.date) return false
        return true
    }

    override fun hashCode(): Int {
        return date.hashCode()
    }
}